### ISSUE TITLE
Have to run macro twice?

### ISSUE THREAD

**Issue by hrobertson2** [GRIFFINCollaboration/detectorSimulations_v10#82]
Hello,

I am running into an issue where I am not able to run my macro directly from the terminal (i.e. by running ./Griffinv10 macro.mac).  If I open it in the GUI and run it once (using /control/execute in the command line), it does not work.  But, if I run it the same way a second time in the GUI, it does work fine.  

I can't seem to explain this behavior as some macros do work on the first run and I can't find an obvious difference that would be causing this in my macro.  I have the same "/vis/" section as the successful macros.  Do you have any thoughts on why this could be?

Thank you!
Heather

**Comment by jsmallcombe** [comment]
It sounds like something (which becomes set during the execution of the macro) needs to be set before running macro.

When you say "not able to run my macro directly from the terminal" what exactly happens? Can you share the full terminal output and also the .mac file.

**Comment by VinzenzBildstein** [comment]
In addition to what James said, could you also explain what "it does not work" means when executing the macro in the GUI?

Like James said, this is most likely something about the order of the commands in the macro, e.g. some of the commands have to come before (or after) the ```/run/initialize``` command.

**Comment by VinzenzBildstein** [comment]
I can't really see in the screenshot you attached what the messages in the terminal are.

What happens if you run the macro w/o the visualization? Try removing the ```/control/execute vis.mac``` line and the following lines to set up the visualization. 

I'm saying that because a) the vis.mac (assuming you use the one provided by the repository?) includes a line ```/run/initialize``` which you already have in your macro, and b) it includes some other lines that you call later on (like ```/vis/open OGL ...```). This might confuse the program since you are initializing the run multiple times and are opening multiple visualization instances.

**Comment by hrobertson2** [comment]
There were never any errors in the terminal window as far as I could tell.

I commented out the `/control/execute vis.mac` and all of the following vis commands and this was the result (Screenshot shows the terminal output at the end of the simulation).

![Screenshot from 2023-03-14 12-06-48](https://user-images.githubusercontent.com/125400808/225111338-00d74f67-32cf-4d50-803d-9acb1648116c.png)

However, the ntuple was still empty when I opened it in GRSISort.

**Comment by VinzenzBildstein** [comment]
When you say the ntuple was empty, do you mean the g4out.root file? Did you also check the g4out_t0.root file?

It also seems that this way your macro runs the first time you start it, so the issue you are having with needing to run it twice seem to stem from the visualization macro. You could play around with it by commenting out specific lines (like the run initialization) and if that works, slowly add the other visualization settings again to see at what point this stops working.

**Comment by hrobertson2** [comment]
Yes, the g4out.root file was empty and I don't seem to have a g4out_t0.root file.  I'll experiment with the commands and see what happens.

Thank you!

**Comment by jsmallcombe** [comment]
Maybe I'm missing something, but it looks like you have created a target and no detectors?
So the ntuple should be empty as there are no "hits" to record.
Or has that changed.

**Comment by hrobertson2** [comment]
I did add some code to DetectorConstruction so that I could detect events on my genericTarget.  When I run the macro a second time within the GUI I am able to see the expected hits.  
A visualization problem seems most likely.  I am working on playing around with the vis commands and will let you know if I get anywhere.

Thank you!

**Comment by VinzenzBildstein** [comment]
In general I would recommend using one of the example run-macros provided, check that it works with that one, and then modify it slowly towards what you want, testing along the way that the results are still what you expect.

**Comment by hrobertson2** [comment]
Thanks Vinzenz!  I'm beginning to suspect that it probably has something to do with the /DetSys/app/genTargetDimensionsCyl command.  I created a new source file based off of the ApparatusGenericTarget.cc file called ApparatusGenericTargetCyl.cc so that I could create a generic cylindrical detector of any material.  When I use the rectangular generic detector it seems to work fine, but the issues arise when I try to use the cylindrical generic detector. (I've attached that file as a .txt and .pdf)

[ApparatusGenericTargetCyl.cc.pdf](https://github.com/GRIFFINCollaboration/detectorSimulations_v10/files/10995687/ApparatusGenericTargetCyl.cc.pdf)
[ApparatusGenericTargetCyl.txt](https://github.com/GRIFFINCollaboration/detectorSimulations_v10/files/10995666/ApparatusGenericTargetCyl.txt)

I can't spot any obvious issues that would be causing this problem, but I also can't find any clear issues in the macro itself so I am a bit stumped.
